azure/services/vnetpeerings/client.go

```diff
@@ -61,29 +61,21 @@ func newPeeringsClient(subscriptionID string, baseURI string, authorizer autores
 }
 
 // Get gets the specified virtual network peering by the peering name, virtual network, and resource group.
-func (ac *AzureClient) Get(ctx context.Context, resourceGroupName, vnetName, peeringName string) (network.VirtualNetworkPeering, error) {
+func (ac *AzureClient) Get(ctx context.Context, spec azure.ResourceSpecGetter) (interface{}, error) {
 	ctx, span := tele.Tracer().Start(ctx, "vnetpeerings.AzureClient.Get")
 	defer span.End()
 
-	return ac.peerings.Get(ctx, resourceGroupName, vnetName, peeringName)
+	return ac.peerings.Get(ctx, spec.ResourceGroupName(), spec.OwnerResourceName(), spec.ResourceGroupName())
 }
 
 // CreateOrUpdateAsync creates or updates a virtual network peering asynchronously.
 // It sends a PUT request to Azure and if accepted without error, the func will return a Future which can be used to track the ongoing
 // progress of the operation.
-func (ac *AzureClient) CreateOrUpdateAsync(ctx context.Context, spec azure.ResourceSpecGetter) (interface{}, azureautorest.FutureAPI, error) {
+func (ac *AzureClient) CreateOrUpdateAsync(ctx context.Context, spec azure.ResourceSpecGetter, existingResource interface{}) (interface{}, azureautorest.FutureAPI, error) {
 	ctx, span := tele.Tracer().Start(ctx, "vnetpeerings.AzureClient.CreateOrUpdateAsync")
 	defer span.End()
 
-	var existingPeering interface{}
-
-	if existing, err := ac.Get(ctx, spec.ResourceGroupName(), spec.OwnerResourceName(), spec.ResourceName()); err != nil && !azure.ResourceNotFound(err) {
-		return nil, nil, errors.Wrapf(err, "failed to get virtual network peering %s for %s in %s", spec.ResourceName(), spec.OwnerResourceName(), spec.ResourceGroupName())
-	} else if err == nil {
-		existingPeering = existing
-	}
-
-	params, err := spec.Parameters(existingPeering)
+	params, err := spec.Parameters(existingResource)
```

Comment:
I wonder if we could go even further and move this logic up in async.go too, then `CreateOrUpdateAsync` can become ` CreateOrUpdateAsync(ctx context.Context, spec azure.ResourceSpecGetter, parameters interface{})`

Reply:
Would we be able to do the type check for params, i.e. `natRule, ok := params.(network.InboundNatRule)` if we move this part into the interface?

Reply:
no that part would stay in here but you could do the nil check in there, ie.:

async.go:

```
if existing, err := client.Get(ctx, spec); err != nil && !azure.ResourceNotFound(err) {
		return nil, errors.Wrapf(err, "failed to get existing resource %s/%s (service: %s)", rgName, resourceName, serviceName)
	} else if err == nil {
		existingResource = existing
	}

params, err := spec.Parameters(existingResource)
if err != nil {
		return nil, nil, errors.Wrapf(err, "failed to get desired parameters for resource %s/%s (service: %s)", rgName, resourceName, serviceName)
} else if params == nil {
   // nothing to do, return existing, don't call CreateOrUpdateAsync
}
```

and in here:

```
func (ac *AzureClient) CreateOrUpdateAsync(ctx context.Context, spec azure.ResourceSpecGetter, parameters interface{}) (interface{}, azureautorest.FutureAPI, error) {
...
peering, ok := parameters.(network.VirtualNetworkPeering)
if !ok {
    return nil, nil, errors.Errorf("%T is not a network.VirtualNetworkPeering", params)
}

future, err := ac.peerings.CreateOrUpdate(ctx, spec.ResourceGroupName(), spec.OwnerResourceName(), spec.ResourceName(), peering, network.SyncRemoteAddressSpaceTrue)

...
```